promql/engine.go

```diff
@@ -919,12 +929,14 @@ func (ev *evaluator) rangeEval(f func([]parser.Value, *EvalNodeHelper) Vector, e
 			for si, series := range matrixes[i] {
 				for _, point := range series.Points {
 					if point.T == ts {
+						ev.samplesStats.Increment(1)
 						if ev.currentSamples < ev.maxSamples {
 							vectors[i] = append(vectors[i], Sample{Metric: series.Metric, Point: point})
 							// Move input vectors forward so we don't have to re-scan the same
 							// past points at the next step.
 							matrixes[i][si].Points = series.Points[1:]
 							ev.currentSamples++
+							ev.samplesStats.UpdatePeak(ev.currentSamples)
```

Comment:
This is an inner loop, you can move this out. Also the stats increment.

Reply:
Sure, assuming you meant outer by one-loop level.

Reply:
The increment I think you can only really move out one, the other can be moved out two - but as long as we're avoiding the inner loop we should be okay. We should check the benchmarks don't show much difference when this is all done.